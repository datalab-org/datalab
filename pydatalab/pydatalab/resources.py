DIRECTORIES = [
    {"name": "bob/Josh_B", "path": "/Volumes/greygroup-instruments/bob/Josh_B"},
    {
        "name": "carlos/Josh_Bocarsly",
        "path": "/Volumes/greygroup-instruments/carlos/Josh_Bocarsly",
    },
    {
        "name": "eve/Josh_Bocarsly",
        "path": "/Volumes/greygroup-instruments/eve/Josh_Bocarsly/",
    },
    {"name": "carlos/Jiasi_Li", "path": "/Volumes/greygroup-instruments/carlos/Jiasi_Li"},
]

DIRECTORIES_DICT = {d["name"]: d for d in DIRECTORIES}
