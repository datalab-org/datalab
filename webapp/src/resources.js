// Resources for the application
import DataBlockBase from "@/components/datablocks/DataBlockBase";
import MediaBlock from "@/components/datablocks/MediaBlock";
import XRDBlock from "@/components/datablocks/XRDBlock";
import ChatBlock from "@/components/datablocks/ChatBlock";
import RamanBlock from "@/components/datablocks/RamanBlock";
import CycleBlock from "@/components/datablocks/CycleBlock";
import NMRBlock from "@/components/datablocks/NMRBlock";
import EISBlock from "@/components/datablocks/EISBlock";
import MassSpecBlock from "@/components/datablocks/MassSpecBlock";

import SampleInformation from "@/components/SampleInformation";
import StartingMaterialInformation from "@/components/StartingMaterialInformation";
import CellInformation from "@/components/CellInformation";
import CollectionInformation from "@/components/CollectionInformation";
import EquipmentInformation from "@/components/EquipmentInformation";

import SampleCreateModalAddon from "@/components/itemCreateModalAddons/SampleCreateModalAddon";
import CellCreateModalAddon from "@/components/itemCreateModalAddons/CellCreateModalAddon";

// Look for values set in .env file. Use defaults if `null` is not explicitly handled elsewhere in the code.
export const API_URL =
  process.env.VUE_APP_API_URL != null ? process.env.VUE_APP_API_URL : "http://localhost:5001";
export const API_TOKEN = process.env.VUE_APP_API_TOKEN;

export const LOGO_URL = process.env.VUE_APP_LOGO_URL;
export const HOMEPAGE_URL = process.env.VUE_APP_HOMEPAGE_URL;

export const GRAVATAR_STYLE = "identicon";

// determine whether inventory should be readonly (except blocks). Note: environment
// variables can only be strings, not bools.
const editable_inventory = process.env.VUE_APP_EDITABLE_INVENTORY || "false";
export const EDITABLE_INVENTORY = editable_inventory.toLowerCase() == "true";

export const UPPY_MAX_TOTAL_FILE_SIZE =
  Number(process.env.VUE_APP_UPPY_MAX_TOTAL_FILE_SIZE) != null
    ? process.env.VUE_APP_UPPY_MAX_TOTAL_FILE_SIZE
    : 102400000000;
export const UPPY_MAX_NUMBER_OF_FILES =
  Number(process.env.VUE_APP_UPPY_MAX_NUMBER_OF_FILES) != null
    ? process.env.VUE_APP_UPPY_MAX_NUMBER_OF_FILES
    : 10000;

export const debounceTime = 250; // time after user stops typing before request is sent

export const blockTypes = {
  comment: { description: "Comment", component: DataBlockBase },
  media: { description: "Media", component: MediaBlock },
  xrd: { description: "Powder XRD", component: XRDBlock },
  raman: { description: "Raman", component: RamanBlock },
  cycle: { description: "Electrochemistry", component: CycleBlock },
  eis: { description: "EIS", component: EISBlock },
  nmr: { description: "NMR", component: NMRBlock },
  ms: { description: "Mass spectrometry", component: MassSpecBlock },
  chat: { description: "Virtual assistant", component: ChatBlock },
};

export const itemTypes = {
  samples: {
    itemInformationComponent: SampleInformation,
    itemCreateModalAddon: SampleCreateModalAddon,
    navbarColor: "#0b6093",
    navbarName: "Sample",
    lightColor: "#d0ebfb",
    labelColor: "#0b6093",
    isCreateable: true,
    display: "sample",
  },
  starting_materials: {
    itemInformationComponent: StartingMaterialInformation,
    navbarColor: "#349579",
    navbarName: "Starting Material",
    lightColor: "#d9f2eb",
    labelColor: "#298651",
    isCreateable: true,
    display: "starting material",
  },
  cells: {
    itemInformationComponent: CellInformation,
    itemCreateModalAddon: CellCreateModalAddon,
    navbarColor: "#946807",
    navbarName: "Cell",
    lightColor: "#D1C28F",
    labelColor: "#946807",
    isCreateable: true,
    display: "cell",
  },
  collections: {
    itemInformationComponent: CollectionInformation,
    navbarColor: "#563D7c",
    navbarName: "Collection",
    lightColor: "#cbd6f7",
    labelColor: "#563D7c",
    display: "collection",
  },
  users: {
    navbarColor: "mediumseagreen",
    navbarName: "User",
    lightColor: "mediumseagreen",
    labelColor: "mediumseagreen",
    isCreateable: false,
    display: "user",
  },
  equipment: {
    itemInformationComponent: EquipmentInformation,
    navbarColor: "#c77c02",
    navbarName: "Equipment",
    lightColor: "#f7d6a1",
    labelColor: "#c77c02",
    display: "equipment",
  },
};

export const SAMPLE_TABLE_TYPES = ["samples", "cells"];
export const INVENTORY_TABLE_TYPES = ["starting_materials"];
export const EQUIPMENT_TABLE_TYPES = ["equipment"];

export const cellFormats = {
  coin: "coin",
  pouch: "pouch",
  in_situ_xrd: "in situ (XRD)",
  in_situ_nmr: "in situ (NMR)",
  in_situ_squid: "in situ (SQUID)",
  swagelok: "swagelok",
  cylindrical: "cylindrical",
  other: "other",
};
