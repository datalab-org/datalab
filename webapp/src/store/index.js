import { createStore } from "vuex";
// import { createLogger } from "vuex";
// import { set } from 'vue'

export default createStore({
  state: {
    all_item_data: {}, // keys: item_ids, vals: objects containing all data
    all_block_data: {},
    sample_list: [],
    starting_material_list: [],
    saved_status: {},
    updating: {},
    updatingDelayed: {},
    remoteDirectoryTree: {},
    remoteDirectoryTreeSecondsSinceLastUpdate: null,
    files: {},
    itemGraphData: null,
    remoteDirectoryTreeIsLoading: false,
    fileSelectModalIsOpen: false,
  },
  mutations: {
    setSampleList(state, sampleSummaries) {
      // sampleSummaries is an array of json objects summarizing the available samples
      state.sample_list = sampleSummaries;
    },
    setStartingMaterialList(state, startingMaterialSummaries) {
      // startingMaterialSummaries is an array of json objects summarizing the available samples
      state.starting_material_list = startingMaterialSummaries;
    },
    appendToSampleList(state, sampleSummary) {
      // sampleSummary is a json object summarizing the new sample
      state.sample_list.push(sampleSummary);
    },
    deleteFromSampleList(state, sample_summary) {
      const index = state.sample_list.indexOf(sample_summary);
      if (index > -1) {
        state.sample_list.splice(index, 1);
      } else {
        console.log("deleteFromSampleList couldn't find the object");
      }
    },
    createItemData(state, payload) {
      // payload should have the following fields:
      // item_id, item_data
      // Object.assign(state.all_sample_data[payload.item_data], payload.item_data)
      state.all_item_data[payload.item_id] = payload.item_data;
      state.saved_status[payload.item_id] = true;
    },
    updateFiles(state, files_data) {
      // payload should be an object with file ids as key and file data as values
      // Note: this will overwrite any entries with the same file_ids
      Object.assign(state.files, files_data);
    },
    addFileToSample(state, payload) {
      state.all_item_data[payload.item_id].file_ObjectIds.push(payload.file_id);
    },
    removeFileFromSample(state, payload) {
      var file_ids = state.all_item_data[payload.item_id].file_ObjectIds;
      const index = file_ids.indexOf(payload.file_id);
      if (index > -1) {
        file_ids.splice(index, 1);
      }
    },
    setRemoteDirectoryTree(state, remoteDirectoryTree) {
      state.remoteDirectoryTree = remoteDirectoryTree.data;
    },
    setRemoteDirectoryTreeSecondsSinceLastUpdate(state, secondsSinceLastUpdate) {
      state.remoteDirectoryTreeSecondsSinceLastUpdate = secondsSinceLastUpdate;
    },
    addABlock(state, { item_id, new_block_obj, new_block_insert_index }) {
      // payload: item_id, new_block_obj, new_display_order

      // I should actually throw an error if this fails!
      console.assert(
        item_id == new_block_obj.item_id,
        "The block has a different item_id (%s) than the item_id provided to addABlock (%s)",
        item_id,
        new_block_obj.item_id
      );
      console.log(`addABlock called with: ${item_id}, ${new_block_obj}, ${new_block_insert_index}`);
      let new_block_id = new_block_obj.block_id;
      state.all_item_data[item_id]["blocks_obj"][new_block_id] = new_block_obj;
      if (new_block_insert_index) {
        state.all_item_data[item_id]["display_order"].splice(
          new_block_insert_index,
          0,
          new_block_id
        );
      }
      // if new_block_insert_index is None, then block is inserted at the end
      else {
        state.all_item_data[item_id]["display_order"].push(new_block_id);
      }
      state.saved_status[item_id] = false;
    },
    updateBlockData(state, payload) {
      // requires the following fields in payload:
      // item_id, block_id, block_data
      console.log("updating block data with:", payload);
      Object.assign(
        state.all_item_data[payload.item_id]["blocks_obj"][payload.block_id],
        payload.block_data
      );
      state.saved_status[payload.item_id] = false;
    },
    updateItemData(state, payload) {
      //requires the following fields in payload:
      // item_id, block_data
      Object.assign(state.all_item_data[payload.item_id], payload.block_data);
      state.saved_status[payload.item_id] = false;
    },
    setSaved(state, payload) {
      // requires the following fields in payload:
      // item_id, isSaved
      state.saved_status[payload.item_id] = payload.isSaved;
    },
    removeBlockFromDisplay(state, payload) {
      // requires the following fields in payload:
      // item_id, block_id
      var display_order = state.all_item_data[payload.item_id].display_order;
      const index = display_order.indexOf(payload.block_id);
      if (index > -1) {
        display_order.splice(index, 1);
      }
    },
    addFile(state, payload) {
      // requires the following fileds in payload:
      // item_id, filename
      state.all_item_data[payload.item_id].files.push(payload.filename);
    },
    removeFilename(state, payload) {
      // requires the following fields in payload:
      // item_id, filename
      var files = state.all_item_data[payload.item_id].files;
      const index = files.indexOf(payload.filename);
      if (index > -1) {
        files.splice(index, 1);
      }
    },
    swapBlockDisplayOrder(state, payload) {
      // requires the following fields in payload:
      // item_id, index1, index2
      // Swaps index1 and index2 in item_id.display_order
      var display_order = state.all_item_data[payload.item_id].display_order;
      if (payload.index1 < display_order.length && payload.index2 < display_order.length) {
        [display_order[payload.index1], display_order[payload.index2]] = [
          display_order[payload.index2],
          display_order[payload.index1],
        ];
      }
      state.saved_status[payload.item_id] = false;
    },
    setBlockUpdating(state, block_id) {
      state.updating[block_id] = true;
      state.updatingDelayed[block_id] = true;
    },
    async setBlockNotUpdating(state, block_id) {
      state.updating[block_id] = false;
      await new Promise((resolve) => setTimeout(resolve, 500));
      state.updatingDelayed[block_id] = false; // delayed by 0.5 s, helpful for some animations
    },
    setFileSelectModalOpenStatus(state, isOpen) {
      state.fileSelectModalIsOpen = isOpen;
    },
    setRemoteDirectoryTreeIsLoading(state, isLoading) {
      state.remoteDirectoryTreeIsLoading = isLoading;
    },
    setItemGraph(state, payload) {
      state.itemGraphData = payload;
    },
  },
  getters: {
    getItem: (state) => (item_id) => {
      return state.all_item_data[item_id];
    },
    getBlockByItemIDandBlockID: (state) => (item_id, block_id) => {
      console.log("getBlockBySampleIDandBlockID called with:", item_id, block_id);
      return state.all_sample_data[item_id]["blocks_obj"][block_id];
    },
  },
  actions: {},
  modules: {},
  // plugins: [createLogger()],
});
