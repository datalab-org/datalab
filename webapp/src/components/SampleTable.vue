<template>
  <div v-if="isSampleFetchError" class="alert alert-danger">
    <font-awesome-icon icon="exclamation-circle" />&nbsp;Server Error. Sample list could not be
    retreived.
  </div>
  <table class="table table-hover table-sm" data-testid="sample-table">
    <thead>
      <tr>
        <th scope="col">ID</th>
        <th scope="col">Type</th>
        <th scope="col">Sample name</th>
        <th scope="col">Formula</th>
        <th class="text-center" scope="col">Date</th>
        <th scope="col">Collections</th>
        <th class="text-center" scope="col">Creators</th>
        <th class="text-center" scope="col"># of blocks</th>
        <th scope="col"></th>
      </tr>
    </thead>
    <tbody>
      <tr
        :id="sample.item_id"
        v-for="sample in samples"
        :key="sample.item_id"
        v-on:click.exact="goToEditPage(sample.item_id)"
        v-on:click.meta="openEditPageInNewTab(sample.item_id)"
        v-on:click.ctrl="openEditPageInNewTab(sample.item_id)"
      >
        <td align="left" class="table-item-id">
          <FormattedItemName
            :item_id="sample.item_id"
            :itemType="sample?.type"
            enableModifiedClick
          />
        </td>
        <td align="center">{{ itemTypes[sample.type].display }}</td>
        <td align="left">{{ sample.name }}</td>
        <td><ChemicalFormula :formula="sample.chemform" /></td>
        <td class="text-center">{{ $filters.IsoDatetimeToDate(sample.date) }}</td>
        <td><CollectionList :collections="sample.collections" /></td>
        <td align="center"><Creators :creators="sample.creators" /></td>
        <td class="text-right">{{ sample.nblocks }}</td>
        <td align="right">
          <button
            type="button"
            class="close"
            @click.stop="deleteSample(sample)"
            aria-label="delete"
          >
            <span aria-hidden="true" style="color: grey">&times;</span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
import ChemicalFormula from "@/components/ChemicalFormula";
import FormattedItemName from "@/components/FormattedItemName";
import CollectionList from "@/components/CollectionList";
import Creators from "@/components/Creators";
import { getSampleList, deleteSample } from "@/server_fetch_utils.js";
import { itemTypes } from "@/resources.js";

export default {
  data() {
    return {
      isSampleFetchError: false,
      itemTypes: itemTypes,
    };
  },
  computed: {
    samples() {
      return this.$store.state.sample_list;
    },
  },
  methods: {
    goToEditPage(item_id) {
      this.$router.push(`/edit/${item_id}`);
    },
    openEditPageInNewTab(item_id) {
      window.open(`/edit/${item_id}`, "_blank");
    },
    // should also check response.OK? And retry if
    getSamples() {
      getSampleList().catch(() => {
        this.isSampleFetchError = true;
      });
    },
    deleteSample(sample) {
      if (confirm(`Are you sure you want to delete sample "${sample.item_id}"?`)) {
        console.log("deleting...");
        deleteSample(sample.item_id);
      }
      console.log("delete cancelled...");
    },
  },
  created() {
    this.getSamples();
  },
  components: {
    ChemicalFormula,
    FormattedItemName,
    Creators,
    CollectionList,
  },
};
</script>

<style scoped>
.table-item-id {
  font-size: 1.2em;
  font-weight: normal;
}
</style>
