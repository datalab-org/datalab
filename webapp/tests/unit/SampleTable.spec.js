import SampleTable from "@/components/SampleTable";
import { render } from "@testing-library/vue";
import store from "@/store/index";
import { API_URL } from "@/resources.js";

beforeEach(() => {
  fetch.resetMocks();
});

var globalPropertiesMock = {
  $API_URL: API_URL,
  $filters: {
    IsoDatetimeToDate(isodatetime) {
      if (isodatetime) {
        return isodatetime.substring(0, 10);
      }
      return isodatetime; // if isodatetime is null or empty, don't do anything
    },
  },
};

test("it renders correctly with 0 samples", () => {
  fetch.mockResponseOnce(
    JSON.stringify({
      samples: [],
    })
  );

  const { getByText, container } = render(SampleTable, {
    global: {
      plugins: [store],
      mocks: globalPropertiesMock,
    },
  });

  getByText("Sample name");
  expect(container).toMatchSnapshot();
});

test("it renders correctly with some samples", async () => {
  await fetch.mockResponseOnce(
    JSON.stringify({
      samples: [
        {
          chemform: "In0.333WO3",
          date: "2020-02-17T00:00",
          name: "This is a test sample",
          nblocks: 6,
          item_id: "jdb1-1",
        },
        {
          chemform: "NaCoO2",
          date: "2021-02-19T08:00",
          name: "in situ SQUID test",
          nblocks: 1,
          item_id: "jdb11-3_e1_s2",
        },
        {
          date: "2021-02-24T09:23",
          name: "another in situ SQUID test",
          nblocks: 1,
          item_id: "jdb11-3_e1_s3",
        },
        {
          date: "1954-01-01T22:22",
          name: " ",
          nblocks: 0,
          item_id: "empty_test",
        },
      ],
    })
  );

  const { findByText, container } = render(SampleTable, {
    global: {
      plugins: [store],
      mocks: globalPropertiesMock,
    },
  });
  // console.log(rendered.container)

  await findByText("jdb1-1");
  await findByText("jdb11-3_e1_s2");
  await findByText("jdb11-3_e1_s3");
  await findByText("empty_test");

  await expect(container).toMatchSnapshot();
});
