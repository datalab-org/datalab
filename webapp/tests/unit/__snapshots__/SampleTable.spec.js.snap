// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly with 0 samples 1`] = `
<div>
  
  <!--v-if-->
  <table
    class="table table-hover table-sm"
  >
    <thead>
      <tr
        align="center"
      >
        <th
          scope="col"
        >
          ID
        </th>
        <th
          scope="col"
        >
          Sample name
        </th>
        <th
          scope="col"
        >
          Formula
        </th>
        <th
          scope="col"
        >
          Date
        </th>
        <th
          scope="col"
        >
          Creators
        </th>
        <th
          scope="col"
        >
          # of blocks
        </th>
        <th
          scope="col"
        />
      </tr>
    </thead>
    <tbody>
      
      
    </tbody>
  </table>
  
</div>
`;

exports[`it renders correctly with some samples 1`] = `
<div>
  
  <!--v-if-->
  <table
    class="table table-hover table-sm"
  >
    <thead>
      <tr
        align="center"
      >
        <th
          scope="col"
        >
          ID
        </th>
        <th
          scope="col"
        >
          Sample name
        </th>
        <th
          scope="col"
        >
          Formula
        </th>
        <th
          scope="col"
        >
          Date
        </th>
        <th
          scope="col"
        >
          Creators
        </th>
        <th
          scope="col"
        >
          # of blocks
        </th>
        <th
          scope="col"
        />
      </tr>
    </thead>
    <tbody>
      
      <tr
        id="jdb1-1"
      >
        <td
          align="left"
        >
          jdb1-1
        </td>
        <td
          align="left"
        >
          This is a test sample
        </td>
        <td>
          <span>
            
            <span>
              In
              <sub>
                0.333
              </sub>
            </span>
            <span>
              W
              <!--v-if-->
            </span>
            <span>
              O
              <sub>
                3
              </sub>
            </span>
            
          </span>
        </td>
        <td
          align="center"
        >
          2020-02-17
        </td>
        <td
          align="center"
        >
          
          
        </td>
        <td
          align="right"
        >
          6
        </td>
        <td
          align="right"
        >
          <button
            aria-label="delete"
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
              style="color: grey;"
            >
              ×
            </span>
          </button>
        </td>
      </tr>
      <tr
        id="jdb11-3_e1_s2"
      >
        <td
          align="left"
        >
          jdb11-3_e1_s2
        </td>
        <td
          align="left"
        >
          in situ SQUID test
        </td>
        <td>
          <span>
            
            <span>
              Na
              <!--v-if-->
            </span>
            <span>
              Co
              <!--v-if-->
            </span>
            <span>
              O
              <sub>
                2
              </sub>
            </span>
            
          </span>
        </td>
        <td
          align="center"
        >
          2021-02-19
        </td>
        <td
          align="center"
        >
          
          
        </td>
        <td
          align="right"
        >
          1
        </td>
        <td
          align="right"
        >
          <button
            aria-label="delete"
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
              style="color: grey;"
            >
              ×
            </span>
          </button>
        </td>
      </tr>
      <tr
        id="jdb11-3_e1_s3"
      >
        <td
          align="left"
        >
          jdb11-3_e1_s3
        </td>
        <td
          align="left"
        >
          another in situ SQUID test
        </td>
        <td>
          <span>
            
            <span>
              
              <!--v-if-->
            </span>
            
          </span>
        </td>
        <td
          align="center"
        >
          2021-02-24
        </td>
        <td
          align="center"
        >
          
          
        </td>
        <td
          align="right"
        >
          1
        </td>
        <td
          align="right"
        >
          <button
            aria-label="delete"
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
              style="color: grey;"
            >
              ×
            </span>
          </button>
        </td>
      </tr>
      <tr
        id="empty_test"
      >
        <td
          align="left"
        >
          empty_test
        </td>
        <td
          align="left"
        >
           
        </td>
        <td>
          <span>
            
            <span>
              
              <!--v-if-->
            </span>
            
          </span>
        </td>
        <td
          align="center"
        >
          1954-01-01
        </td>
        <td
          align="center"
        >
          
          
        </td>
        <td
          align="right"
        >
          0
        </td>
        <td
          align="right"
        >
          <button
            aria-label="delete"
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
              style="color: grey;"
            >
              ×
            </span>
          </button>
        </td>
      </tr>
      
    </tbody>
  </table>
  
</div>
`;
